File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        if (!head) {
            return false;
        }

        ListNode* tortoise = head;
        ListNode* hare = head->next;
        while (tortoise && hare && hare->next) {
            if (tortoise == hare) {
                return true;
            }
            tortoise = tortoise->next;
            hare = hare->next->next;
        }
        
        return false;
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        def rec(prev, node, n) -> int:
            if node is None:
                return n
            res = rec(node, node.next, n)
            if res == 1:
                prev.next = node.next
            return res - 1
        dummy = ListNode()
        dummy.next = head
        rec(dummy, head, n)
        return dummy.next